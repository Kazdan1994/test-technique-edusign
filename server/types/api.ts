enum Status {
  Success = "success",
  Error = "error",
}

interface SuccessResponse<T> {
  status: Status.Success;
  result: T;
}

// Define a type for the error response
interface ErrorResponse {
  status: Status.Error;
  message: string;
  errorCode: string;
}

// Define a generic type for either success or error response
type ApiResponse<T> = SuccessResponse<T> | ErrorResponse;

export type { ApiResponse, SuccessResponse, ErrorResponse };
export { Status };
