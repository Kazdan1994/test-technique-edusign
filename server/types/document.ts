interface DocumentData {
  documentsSuccess: number;
  documents: string[];
}

export type { DocumentData };
