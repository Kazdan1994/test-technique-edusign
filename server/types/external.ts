export interface ExternalData {
  ID: string;
  SCHOOL_ID: string;
  FIRSTNAME: string;
  LASTNAME: string;
  EMAIL: string;
  COMPANY: string;
  PHONE: string;
  DATE_CREATED: string | null;
  DATE_UPDATED: string | null;
  TYPE: number;
  API_ID: string;
  API_TYPE: string;
  HIDDEN: number;
}
