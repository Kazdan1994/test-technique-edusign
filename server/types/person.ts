export type Person = {
  id: string;
  email: string;
};
