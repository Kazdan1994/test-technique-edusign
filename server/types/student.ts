export interface StudentData {
  ID: string;
  id: string;
  EMAIL: string;
  FIRSTNAME: string;
  LASTNAME: string;
  USERNAME: string;
  PHOTO: null | string;
  PHONE: string;
  TRAINING_NAME: string;
  FILE_NUMBER: null | number;
  COMPANY: string;
  TAGS: string[];
  SIGNATURE_ID: string;
  HIDDEN: number;
  DATE_CREATED: string | null;
  DATE_UPDATED: string | null;
  MULTI_ACCOUNT_LOGIN_CODE: number;
  SCHOOL_ID: string;
  BADGE_ID: string;
  API_ID: string;
  API_TYPE: string;
  LANGUAGE: string;
  NEW_PASSWORD_NEEDED: number;
  STUDENT_FOLLOWER_ID: string[];
  GROUPS: string[];
  VARIABLES: string[];
}
