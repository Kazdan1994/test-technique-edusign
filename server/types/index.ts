import type { ApiResponse, SuccessResponse, ErrorResponse } from "./api";
import { Status } from "./api";
import type { StudentData } from "./student";
import type { Person } from "./person";

export type {
  ApiResponse,
  SuccessResponse,
  ErrorResponse,
  StudentData,
  Person,
};

export { Status };
