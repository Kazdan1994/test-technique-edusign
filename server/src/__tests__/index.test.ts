import request from "supertest";
import { faker } from "@faker-js/faker";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { getIntervenants, getStudent, sendDocuments, server } from "..";
import { ApiResponse, ErrorResponse, Status, StudentData } from "../../types";
import { ExternalData } from "../../types/external";

const mock = new MockAdapter(axios);

const studentFirstName = faker.person.firstName();
const studentLastName = faker.person.lastName();
const emailStudent = faker.internet.email({
  firstName: studentFirstName,
  provider: "edusign.fr",
});
const emailsIntervenants = [faker.internet.email()];

describe("app test", () => {
  afterAll(() => {
    server.close();
  });

  afterEach(() => {
    mock.reset();
  });

  test("home route", async () => {
    const res = await request(server).get("/");

    expect(res.statusCode).toEqual(200);
    expect(res.body).toHaveProperty("homeText");
    expect(res.body.homeText).toBe("My super home text");
  });

  test("get student email", async () => {
    const id = faker.string.sample(16);
    const dataResponse: ApiResponse<StudentData> = {
      status: Status.Success,
      result: {
        ID: id,
        id,
        EMAIL: emailStudent,
        FIRSTNAME: studentFirstName,
        LASTNAME: studentLastName,
        USERNAME: faker.internet.userName({
          firstName: studentFirstName,
          lastName: studentLastName,
        }),
        PHOTO: null,
        PHONE: "",
        TRAINING_NAME: "",
        FILE_NUMBER: null,
        COMPANY: "",
        TAGS: [],
        SIGNATURE_ID: "",
        HIDDEN: 0,
        DATE_CREATED: faker.date.recent().toISOString(),
        DATE_UPDATED: null,
        MULTI_ACCOUNT_LOGIN_CODE: 0,
        SCHOOL_ID: faker.string.sample(16),
        BADGE_ID: "",
        API_ID: "",
        API_TYPE: "",
        LANGUAGE: "fr",
        NEW_PASSWORD_NEEDED: 1,
        STUDENT_FOLLOWER_ID: [],
        GROUPS: [],
        VARIABLES: [],
      },
    };

    mock
      .onGet("https://ext.edusign.fr/v1/student/by-email", {
        params: {
          email: emailStudent,
        },
      })
      .reply(200, dataResponse);

    const student = await getStudent(emailStudent);

    expect(student.id).toBe(dataResponse.result.id);
    expect(student.email).toBe(emailStudent);
  });

  test("get error if student doesn't exist", async () => {
    const errorResponse: ApiResponse<ErrorResponse> = {
      status: Status.Error,
      message: "An error occured",
      errorCode: "ERROR",
    };

    const emailStudentDoesNotExist = faker.internet.email();
    mock
      .onGet("https://ext.edusign.fr/v1/student/by-email", {
        params: {
          email: emailStudentDoesNotExist,
        },
      })
      .reply(200, errorResponse);

    await expect(
      getStudent(emailStudentDoesNotExist),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test("get intervenants emails", async () => {
    const emailIntervenant = emailsIntervenants[0];

    const dataResponse: ApiResponse<ExternalData> = {
      status: Status.Success,
      result: {
        ID: faker.string.sample(16),
        SCHOOL_ID: faker.string.sample(16),
        FIRSTNAME: faker.person.firstName(),
        LASTNAME: faker.person.lastName(),
        EMAIL: emailIntervenant,
        COMPANY: "",
        PHONE: "",
        DATE_CREATED: faker.date.recent().toISOString(),
        DATE_UPDATED: null,
        TYPE: 0,
        API_ID: "",
        API_TYPE: "",
        HIDDEN: 0,
      },
    };

    mock
      .onGet("https://ext.edusign.fr/v1/externals/by-email", {
        params: {
          email: emailIntervenant,
        },
      })
      .reply(200, dataResponse);

    const intervenants = await getIntervenants(emailsIntervenants);

    expect(intervenants[0].id).toBe(dataResponse.result.ID);
    expect(intervenants[0].email).toBe(emailIntervenant);
  });

  test("get error if one intervenant email is wrong", async () => {
    const errorResponse: ApiResponse<ErrorResponse> = {
      status: Status.Error,
      message: "ERR_UNKNOWN_EXTERNAL_BY_EMAIL",
      errorCode: "ERROR",
    };
    const emailExternalDoesNotExist = faker.internet.email();

    mock
      .onGet("https://ext.edusign.fr/v1/externals/by-email", {
        params: {
          email: emailExternalDoesNotExist,
        },
      })
      .reply(200, errorResponse);

    await expect(
      getIntervenants([emailExternalDoesNotExist]),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test("send documents to recipients", async () => {
    // const emailsIntervenants = ["sebastien+391948@edusign.fr"];
    const response =
      await sendDocuments(/*[emailStudent, ...emailsIntervenants]*/);

    expect(response.status).toBe(200);
    expect(response).toHaveProperty("message");
    expect(response.message).toBe("ok");
  });
});
