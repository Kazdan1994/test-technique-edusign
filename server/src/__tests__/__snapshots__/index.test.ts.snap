// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app test get error if one intervenant email is wrong 1`] = `"ERR_UNKNOWN_EXTERNAL_BY_EMAIL"`;

exports[`app test get error if student doesn't exist 1`] = `"An error occured"`;
